### DISCUSSION TITLE
Proposal: Simplifying the Creation of Dual Methods (sync and async) in C#

### DISCUSSION THREAD

**Discussion by precisioninfinity** [dotnet/csharplang#7052] [Language Ideas]
Introduction:
Creating dual methods in C# (i.e., methods that can be executed synchronously or asynchronously) can be a clunky and error-prone process, particularly for new developers who are just getting started with the language. This proposal seeks to simplify the process of creating dual methods by introducing a new keyword, asyncable, that will automatically generate synchronous and asynchronous versions of a method at compile time. The reservations of adding a new keyword here hopefully will be mitigated by it being an obvious offshoot of the async keyword, and hopefully it's understood that asyncable means it can take a dual form sync and async in the method that has that keyword.

Benefits:
The proposed changes would provide several benefits for C# developers, including:

Simplified code: The asyncable keyword would make it easier to write code that can be executed synchronously or asynchronously, reducing the amount of boilerplate code required for this task.

Improved readability: The use of the asyncable keyword would make it easier to read and understand code that can be executed synchronously or asynchronously, particularly for new developers who are just getting started with C#.

Reduced learning curve: The proposed changes would make it easier for new developers to learn how to create dual methods in C#, reducing the learning curve for this important task.

Improved compatibility: The proposed changes would be compatible with existing code and frameworks, ensuring that developers can integrate them into their existing projects without introducing new issues or incompatibilities.

Examples:
Here are some examples of how the proposed changes would simplify the process of creating dual methods in C#:
```csharp 
// Example #1 Simplest example of using this
// This would compile down to a Task MyMethodAsync() and a void MyMethod()
// if there's already either a void MyMethod() or an async Task MyMethodAsync then there'd be a compile time error
public asyncable void MyMethod()
{
    // this line would compile differently in the two methods, one would be
    // as written sync method with the awaitable keyword removed,
    // the other would be compiled down to an 'await client.GetAsync(...)' method
    // You'd get a compile time error if the method you're calling here does not have an
    // async method with GetAsync syntax and the same params
    awaitable client.Get("https://example.com"); // compiles separately to async I/O operation and sync io operation
}


// Best equivalent I can think of doing this currently without using the
// proposed asyncable/awaitable keywords above requires writing 3 methods 
// but usually it's a lot messier, harder to read, and even more challenging than this for new devs
// By comparing these three methods below to the one method above, you can see how much cleaner
// the proposed idea above is
public ValueTask<MyResult> MyMethodAsync()
{
    return MyMethodInternalAsync(false);
}

public MyResult MyMethod()
{
    return MyMethodInternalAsync(true).Result;
}

private async ValueTask<MyResult> MyMethodInternalAsync(bool runSynchronously)
{
    using (var client = new HttpClient())
    {
        if (runSynchronously)
        {
            // perform synchronous operation
            return new MyResult();
        }
        else
        {
            // perform asynchronous operation
            var result = await client.GetAsync("https://example.com");
            // process the result asynchronously
            return new MyResult();
        }
    }
}


// Example #2 Alternate example that gives more control over how the async vs sync operation happens
public asyncable void MyMethod2()
{
     // the awaitable keyword doubles as a boolean value. 
    if (awaitable)
    {
        using (var client = new HttpClient())
        {
            // await keyword can only be used in an if (asyncable) checked block
            var result = await client.GetAsync("https://example.com"); 
        }
    }
    else
    {
        using (var client = new HttpClient())
        {
            // But the awaitable keyword can be used anywhere
            // It might better to NOT use it here because this is targeting the sync form, 
            // just showing as an example because it is possible
            var result = awaitable client.Get("https://example.com"); 
        }
    }
}
```

Conclusion:
The proposed changes to the C# language would simplify the process of creating dual methods, making it easier for developers to write, read, and understand code that can be executed synchronously or asynchronously. By reducing the learning curve for this important task, the proposed changes would make it easier for new developers to get started with C# and become productive members of the development community.

PS:
Given a little more thought, this might be tweaked a bit to help with the async zombie virus problem.

**Comment by CyrusNajmabadi** [comment]
i'd be intrigued with what could potentially be done with SGs here.

**Reply by precisioninfinity** [reply]
Yeah I was thinking of that. I think code gen should be the way to go to flush this out first. In fact, it would be nice if the C#Lang Team sanctioned and published SG stuff they recommend for people that want to try out new language feature proposals like this. I think they'd get more feedback for what works than their current system because I'd be more inclined to use a codegen preview of an upcoming feature proposal than a preview build of the language. Not sure sure if they do that.

That being said, there are benefits of having this in the language eventually. A common standard is nicer for starters, (whereas Code Gens would be company and project specific so everyone does their own thing). Compile time feedback might be better. Analyzers and AI generators like copilot would give better suggestions if there was a larger amount of code doing things a standardized way, etc.

**Reply by CyrusNajmabadi** [reply]
> it would be nice if the C#Lang Team sanctioned 

I formally sanction this.

**Reply by CyrusNajmabadi** [reply]
> whereas Code Gens would be company and project specific so everyone does their own thing)

With this idea, we wouldn't have libraries. Everything would just be built into the language. The idea at the language level is to make this stuff possible. Then we can achieve efficiency of scale as now the entire ecosystem can build and provide value, instead of it always being bottlenecked by what a small team can deliver.

**Reply by precisioninfinity** [reply]
Right, it's not that libraries are bad, but the async libraries still are a bit cludgy to use for this scenario. SG is a nice middle ground where the C# team can 'extend' the C# language without actually making changes to C#.  Then if an SG "extension or plugin" to C# gets a lot of popularity or traction or use, they can consider rolling it into the language. The Visual Studio team does something similar by trying out stuff in extensions and then rolling them into visual studio. Same concept, but if the C# Lang team published 'extensions' to C#, I'd be willing to use that in my production code whereas I would never consider using a preview of the language. One more advantage, the C# Lang SG Extension could be BW Compatible to previous versions of the language.

**Reply by CyrusNajmabadi** [reply]
Would definitely recommend something for Roslyn or roslyn-analyzers to consider :-)

I will formally sanction that as well on the part of the land design group :-)

**Comment by HaloFour** [comment]
This is something that could be achieved with source generators, in that the developer writes, say, the sync version of the code and the source generator emits the async version.  However, that does require that the analyzer (or compiler) can understand the async equivalents of any given synchronous method, assuming one exists (or vice versa) which makes such automatic solutions somewhat brittle in general.

**Reply by CyrusNajmabadi** [reply]
seems somewhat sane that, outside of some small exceptions, you'd just look for a method with the same name (with teh 'Async' suffix) and params, and an 'awaitable' return type of hte original return type.  I think that would go very far with being sufficient.

**Reply by HaloFour** [reply]
Agreed, also include a `CancellationToken` is the signature and include that in the async calls if it is exposed in those signatures.  With source attributes you could better inform the generator as to which invocations are expected to have async equivalents so it could warn if it couldn't figure out the async version.

**Reply by CyrusNajmabadi** [reply]
Great point on CancellationTOken :)

I agrre that the most ideal would be an explicit marking, so there was no confusion.  But absent that, it feels pretty reasonable  to figure out.  And, in the rare case it doesn't, you write it manually :)
